filestore/fsrefstore.go

```diff
@@ -0,0 +1,199 @@
+package filestore
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/ipfs/go-ipfs/blocks"
+	"github.com/ipfs/go-ipfs/blocks/blockstore"
+	pb "github.com/ipfs/go-ipfs/filestore/pb"
+	dshelp "github.com/ipfs/go-ipfs/thirdparty/ds-help"
+	posinfo "github.com/ipfs/go-ipfs/thirdparty/posinfo"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	dsns "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/namespace"
+	dsq "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/query"
+	proto "gx/ipfs/QmT6n4mspWYEya864BhCUJEgyxiRfmiSY9ruQwTUNpRKaM/protobuf/proto"
+	cid "gx/ipfs/QmcTcsTvfaeEBRFo1TkFgT8sRmgi1n1LTZpecfVP8fzpGD/go-cid"
+)
+
+var FilestorePrefix = ds.NewKey("filestore")
+
+type FileManager struct {
+	ds   ds.Batching
```

Comment:
Note.  In my early implementations I wrote the filestore on top of the datastore also.  However, I changed this to directly use the leveldb as I found little value in the abstraction and extra layers.  What actually prompted it was https://github.com/ipfs-filestore/go-ipfs/issues/10.  In particular querying the datastore was very very slow, however I then latter find it very useful to use the features of the LevelDB in order to provide safe verification of the filestore while the deamon was running by using the LevelDB snapshot feature.  With my recent improvements to the datastore query the performance may no longer be a problem, but I still see little value in the extra indirection.